GoogleApiWrappers/src/Google.Apis.Storage.v1.ClientWrapper/StorageClient.DownloadObject.cs

```diff
@@ -0,0 +1,165 @@
+        }
+
+        public Task DownloadObjectAsync(string bucket, string sourceName, Stream destination)
+        {
+            return DownloadObjectAsync(bucket, sourceName, destination,
+                options: null, cancellationToken: CancellationToken.None, progress: null);
+        }
+
+        public Task DownloadObjectAsync(
+            string bucket,
+            string sourceName,
+            Stream destination,
+            DownloadObjectOptions options,
+            CancellationToken cancellationToken,
+            IProgress<IDownloadProgress> progress)
+        {
+            string uri = GetUri(bucket, sourceName);
+            return DownloadObjectAsyncImpl(uri, destination, options, cancellationToken, progress);
+        }
+
+        public Task DownloadObjectAsync(Object source, Stream destination)
+        {
+            return DownloadObjectAsync(source, destination,
+                options: null, cancellationToken: CancellationToken.None, progress: null);
+        }
+
+        public void DownloadObject(Object source, Stream destination)
+        {
+            DownloadObject(source, destination, options: null, progress: null);
+        }
+
+        public void DownloadObject(
+            Object source,
+            Stream destination,
+            DownloadObjectOptions options,
+            IProgress<IDownloadProgress> progress)
+        {
+            string uri = GetUri(source);
+            DownloadObjectImpl(uri, destination, options, progress);
+        }
+
+        public Task DownloadObjectAsync(
+            Object source,
+            Stream destination,
+            DownloadObjectOptions options,
+            CancellationToken cancellationToken,
+            IProgress<IDownloadProgress> progress)
+        {
+            string uri = GetUri(source);
+            return DownloadObjectAsyncImpl(uri, destination, options, cancellationToken, progress);
+        }
+
+        /// <summary>
+        /// Constructs the media URL of an object from its bucket and name.
+        /// </summary>
+        private static string GetUri(string bucket, string sourceName)
+        {
+            bucket.CheckNotNull(nameof(bucket));
+            sourceName.CheckNotNull(nameof(sourceName));
+            return $"https://www.googleapis.com/download/storage/v1/b/{Uri.EscapeDataString(bucket)}/o/{Uri.EscapeDataString(sourceName)}?alt=media";
+        }
+
+        /// <summary>
+        /// Obtains the media URL of an object from eiteher the MediaLink property, or its bucket and name.
+        /// </summary>
+        private static string GetUri(Object source)
+        {
+            source.CheckNotNull(nameof(source));
+            if (source.MediaLink != null)
+            {
+                return source.MediaLink;
+            }
+            if (source.Bucket != null && source.Name != null)
+            {
+                return GetUri(source.Bucket, source.Name);
```

Comment:
There should always be a media link if the Object was generated by the service. That said, you may want to consider adding support for pinning to a specific generation, since if the object is overwritten between calls, you may not actually want the new version since your logic about whether to download may have been made based on the old version.